src/snowflake/snowpark/session.py

```diff
@@ -1638,46 +1700,47 @@ def _resolve_packages(
         #  'python-dateutil': 'python-dateutil==2.8.2'}
         # Add to packages dictionary. Make a copy of existing packages
         # dictionary to avoid modifying it during intermediate steps.
-        result_dict = (
-            existing_packages_dict.copy() if existing_packages_dict is not None else {}
-        )
-
-        # Retrieve list of dependencies that need to be added
-        dependency_packages = self._get_dependency_packages(
-            package_dict,
-            validate_package,
-            package_table,
-            result_dict,
-            statement_params=statement_params,
-        )
+        with self._package_lock:
+            result_dict = (
+                existing_packages_dict if existing_packages_dict is not None else {}
```

Comment:
why the .copy() is removed here?

Reply:
The original version of this code had this pattern.

```
result_dict = existing_packages_dict if existing_pacakges_dict else {}
...
for package in dependency_packages:
    name = package.name
    if name in result_dict:
        ...
    else:
        result["name"] = name
```

I refactored it the wrong way using this pattern
```
result_dict = existing_packages_dict.copy() if existing_pacakges_dict else {}
...
for package in dependency_packages:
    name = package.name
    if name in result_dict:
        ...
    else:
        result["name"] = name

if existing_packages_dict:
    existing_pacakges_dict.update(result_dict)
```
I noticed this refactor is not correct as the update step can overwrite updates from a different thread.

So in this change, I undid the refactor and updated the code as follows:
```
with self._package_lock:
    result_dict = existing_packages_dict if existing_pacakges_dict else {}
    ...
    for package in dependency_packages:
        name = package.name
        if name in result_dict:
            ...
        else:
            result["name"] = name
```